added Alternative Domain section to dns page in docs (#11458)
hashicorp/consul: website/content/docs/discovery/dns.mdx

Alternatively, the TXT record will only include the node's metadata value when the
node's metadata key starts with `rfc1035-`.

## Alternative Domain

> Consul support setting an alternative domain, which is usefull during DNS migrations or if you want to response

Reviewer: ```suggestion
Consul support setting an alternative domain, which is usefull during DNS migrations or if you want to respond
```

> Consul support setting an alternative domain, which is usefull during DNS migrations or if you want to respond to
> internal and external queries with different domains.
> 
> If you are using [`alt_domain`](/docs/agent/options#alt_domain) parameter, Consul will respond  based on the presence of it in your queries.

Reviewer 2: I liked your use case examples! I made some changes to order and wording. Let me know what you think.

```suggestion
By default, Consul responds to DNS queries only for its configured
[`domain`](/docs/agent/options#domain).

Some use cases require responding to queries for more than one domain,
such as during a DNS migration or to distinguish between internal and
external queries by using different domains.

Consul can configured to respond to DNS queries on an alternative domain
through the [`alt_domain`](/docs/agent/options#alt_domain) agent configuration
option. Consul's DNS response will use the same domain as was used in the query.
```

> For exmaple if `test-domain` is configured as an alternative domain the following query:

Reviewer 2: ```suggestion
For example, if `test-domain` is configured as the alternative domain, the following query:
```

```shell-session
$ dig @127.0.0.1 -p 8600  consul.service.test-domain SRV
```
will return:

```
;; QUESTION SECTION:
;consul.service.test-domain.	IN	SRV

;; ANSWER SECTION:
consul.service.test-domain. 0	IN	SRV	1 1 8300 machine.node.dc1.test-domain.

;; ADDITIONAL SECTION:
machine.node.dc1.test-domain. 0	IN	A	127.0.0.1
machine.node.dc1.test-domain. 0	IN	TXT	"consul-network-segment="
```

> -> **Note:** Response to `<ip>.in-addr.arpa.` will always be returned with your default domain, as there is no way to identify queried domain.

Reviewer 2: ```suggestion
-> **Note:** Responses to PTR queries (`<ip>.in-addr.arpa.`) will always use the
[primary domain](/docs/agent/options#domain) (not the alternative domain),
as there is no way for the query to specify a domain.
```

Alternatively, the TXT record will only include the node's metadata value when the
node's metadata key starts with `rfc1035-`.

## Alternative Domain

By default, Consul responds to DNS queries only for its configured
[`domain`](/docs/agent/options#domain).

Some use cases require responding to queries for more than one domain,
such as during a DNS migration or to distinguish between internal and
external queries by using different domains.

> Consul can be configured to respond to DNS queries on an alternative domain

Reviewer 2: We should indicate in which versions the `alt_domain` option is available.

```suggestion
Consul versions 1.5.2+ can be configured to respond to DNS queries on an alternative domain
```

through the [`alt_domain`](/docs/agent/options#alt_domain) agent configuration

> option. Consul's DNS response will use the same domain as was used in the query.

Reviewer 2: I just realized we don't indicate which version(s) this statement is true for. @kbabuadze, @trujillo-adam : what do you think of this?

```suggestion
option. As of Consul versions 1.11.0+, Consul's DNS response will use the same domain as was used in the query;
in prior versions, the response may use the primary [`domain`](/docs/agent/options#domain) no matter which
domain was used in the query.
```

Reviewer 2: @trujillo-adam : an idea for you to consider longer term (outside the scope of this PR) ... should sections like this have a version compatibility table? Something like the below... It could allow us to be more explicit about where support starts and how it changes (e.g., bugfixes, enhancements).

|------------------|------------------------|
| version range   | support                      |
| <1.5.2               | not available              |
| 1.5.2 - 1.10       | Consul DNS will respond to queries on the `alt_domain`, but the response may use the primary domain (e.g., `consul.`)
| 1.11.0+            | Consul DNS will respond to queries on the `alt_domain` with the same domain as was used in the query |

> By default, Consul responds to DNS queries only for its configured
> [`domain`](/docs/agent/options#domain).
> 
> Some use cases require responding to queries for more than one domain,
> such as during a DNS migration or to distinguish between internal and
> external queries by using different domains.
> 
> Consul can be configured to respond to DNS queries on an alternative domain
> through the [`alt_domain`](/docs/agent/options#alt_domain) agent configuration
> option. Consul's DNS response will use the same domain as was used in the query.
> For example, if `test-domain` is configured as the alternative domain, the following query:

Reviewer 3: ```suggestion
By default, Consul responds to DNS queries in the `consul` domain, but you can set a specific domain for responding to DNS queries by configuring the [`domain`](/docs/agent/options#domain) parameter.

You can specify an alternative domain for DNS queries in the `alt_domain` parameter. Configuring the [`alt_domain`](/docs/agent/options#alt_domain) agent parameter enables Consul to respond to DNS queries to the specified domain.

In the following example, the `alt_domain` parameter is set to `test-domain`:

<CodeTabs heading="agent.hcl">
<CodeBlockConfig>

```hcl
alt_domain = "test-domain"
```

</CodeBlockConfig>
<CodeBlockConfig>

```json
"alt_domain" : "test-domain"
```

</CodeBlockConfig>
</CodeTabs>

The configuration enables the following `test-domain` query to run successfully:
```
I was just trying out some alternate wording to see if I could make it smoother. Please ignore if it doesn't seem like an improvement. 
I also wanted to show an example of the `alt_domain` configuration. This is because it wasn't clear from looking at the entry in the reference whether the option took a Boolean value to enable alt domains or if you have to specify the actual domain as a string.

Reviewer 2: My personal opinion is that the new wording leaves out some important information (e.g., why someone might want to use an alternative domain, what domain is used in the response). I personally prefer the previous wording.

However, it is problematic if the `alt_domain` configuration explanation doesn't make its use clear.

@trujillo-adam : Does it make more sense to update the description on the agent configuration option page to make its use more obvious (and perhaps include an example)? Or perhaps provide the configuration example here as you suggested, but also link to this "Alternative Domain" section on the DNS page from the `alt_domain` agent config option description?

Reviewer 3: Ah, that's because I accidentally left a paragraph out that I planned on putting back in at line 102:

"In some instances, Consul may need to respond to queries in more than one domain, such as during a DNS migration or to distinguish between internal and external queries. "

As far as updating the alt_domain reference, yes, we should update it so include at least the data type and default. As far as cross-linking, we should probably think about it a little more. We ca either put 0 guidance in the reference and link internal pages for usage or vice versa. We're currently a little inconsistent. For now I think just adding a little more detail to the reference section is good enough.

Reviewer 2: @kbabuadze - can you add the HCL configuration example @trujillo-adam suggested?

```shell-session
$ dig @127.0.0.1 -p 8600  consul.service.test-domain SRV
```

> will return:

Reviewer 3: ```suggestion
The following responses are returned:
```

```
;; QUESTION SECTION:
;consul.service.test-domain.	IN	SRV

;; ANSWER SECTION:
consul.service.test-domain. 0	IN	SRV	1 1 8300 machine.node.dc1.test-domain.

;; ADDITIONAL SECTION:
machine.node.dc1.test-domain. 0	IN	A	127.0.0.1
machine.node.dc1.test-domain. 0	IN	TXT	"consul-network-segment="
```

> -> **Note:** Responses to PTR queries (`<ip>.in-addr.arpa.`) will always use the

Reviewer 3: ```suggestion
-> **PTR queries:**  Responses to PTR queries (`<ip>.in-addr.arpa.`) will always use the
```
I think it would be a good idea to start using more meaningful labels in these callout blocks. Also, is a PTR query a pointer query? If so, we should say "Responses to pointer (PTR) queries . . ."
